Firestore/core/include/firebase/firestore/document_reference.h

```diff
@@ -0,0 +1,375 @@
+// they exist)
+namespace firebase {
+class App;
+template <typename T>
+class Future;
+}  // namespace firebase
+
+namespace firebase {
+namespace firestore {
+
+// TODO(rsgowman): replace these forward decl's with appropriate includes (once
+// they exist)
+class FieldValue;
+class DocumentSnapshot;
+class Firestore;
+class Error;
+template <typename T>
+class EventListener;
+class ListenerRegistration;
+class CollectionReference;
+class DocumentListenOptions;
+// TODO(rsgowman): not quite a forward decl, but required to make the default
+// parameter to Set() "compile".
+class SetOptions {
+ public:
+  SetOptions();
+};
+
+// TODO(rsgowman): move this into the FieldValue header
+#ifdef STLPORT
+using MapFieldValue = std::tr1::unordered_map<std::string, FieldValue>;
+#else
+using MapFieldValue = std::unordered_map<std::string, FieldValue>;
+#endif
+
+/**
+ * A DocumentReference refers to a document location in a Firestore database and
+ * can be used to write, read, or listen to the location. There may or may not
+ * exist a document at the referenced location. A DocumentReference can also be
+ * used to create a CollectionReference to a subcollection.
+ *
+ * Create a DocumentReference via Firebase::Document(const string& path).
+ *
+ * Subclassing Note: Firestore classes are not meant to be subclassed except for
+ * use in test mocks. Subclassing is not supported in production code and new
+ * SDK releases may break code that does so.
+ */
+class DocumentReference {
+ public:
+  /**
+   * @brief Default constructor. This creates an invalid DocumentReference.
+   * Attempting to perform any operations on this reference will fail (and cause
+   * a crash) unless a valid DocumentReference has been assigned to it.
+   */
+  DocumentReference();
+
+  /**
+   * @brief Copy constructor. It's totally okay (and efficient) to copy
+   * DocumentReference instances, as they simply point to the same location in
+   * the database.
+   *
+   * @param[in] reference DocumentReference to copy from.
+   */
+  DocumentReference(const DocumentReference& reference);
+
+  /**
+   * @brief Move constructor. Moving is an efficient operation for
+   * DocumentReference instances.
+   *
+   * @param[in] reference DocumentReference to move data from.
+   */
+  DocumentReference(DocumentReference&& reference);
+
+  virtual ~DocumentReference();
+
+  /**
+   * @brief Copy assignment operator. It's totally okay (and efficient) to copy
+   * DocumentReference instances, as they simply point to the same location in
+   * the database.
+   *
+   * @param[in] reference DocumentReference to copy from.
+   *
+   * @returns Reference to the destination DocumentReference.
+   */
+  DocumentReference& operator=(const DocumentReference& reference);
+
+  /**
+   * @brief Move assignment operator. Moving is an efficient operation for
+   * DocumentReference instances.
+   *
+   * @param[in] reference DocumentReference to move data from.
```

Comment:
Are these `[in]` suffixes actually used by anything? As it stands this is kind of a lie. It's more of an in-out parameter because whatever implementation was inside this implementation is now swapped with the incoming reference.

Unless they're required the types already say what's going on:
  * `Foo` is always an in parameter
  * `const Foo&` is always an in parameter
  * `Foo&` is an inout parameter (discouraged/prohibited by the style guide, but happens sometimes (e.g. swap))
  * `Foo*` is an inout or maybe just out parameter
  * `Foo&&` is almost always inout

Reply:
It's a doxygen convention. (Doxygen allows this to be [in], [out], or [in,out]). Other firebase header files use it (and either declare the move constructor param as [in] or don't mention the direction at all), though it doesn't seem to actually make it to the external docs.

Some types make it redundant (eg const Foo& as you point out). With others, it's unclear (Foo* could be any of the above, though if 'in' then one would expect to see 'const' next to it, at least in codebases that use const such as ours.)

The direction is conventionally not quite the same as const (though there's obviously nothing enforcing that). 'out' would imply the caller should read that variable after calling the function to get additional information (which they clearly should not do after calling a move ctor.) At least, that's how I've always interpreted it, but I see the doxygen site doesn't actually state this.

I think this is pretty low value. In my personal code, I tend to omit [in] though do include [out] when I can't easily avoid an output parameter. (I've included direction here for consistency with the rest of firebase; though firebase doesn't consistently use it everywhere.)